File: funsets.hs
contains     s x = s x
singletonSet e x = x == e

union     s t x = s x || t x
intersect s t x = s x && t x
diff      s t x = s x && not (t x)
filter          = intersect

bound = 1000

forall s p = null [x | x <- [-bound..bound], s x && not (p x)]
exists s p = not $ forall s (\x -> not $ p x)

map s f y = not $ null [x | x <- [-bound..bound], f x == y && s x]

setToList s = [x | x <- [-bound..bound], s x]
printSet = putStrLn . show . setToList

File: funsets.scala
package funsets
import common._

object FunSets {
  type Set = Int => Boolean
  def contains(s: Set, elem: Int): Boolean = s(elem)
  def singletonSet(elem: Int): Set = (x: Int) => x == elem

  def union(s: Set, t: Set): Set = (x: Int) => s(x) || t(x)
  def intersect(s: Set, t: Set): Set = (x: Int) => s(x) && t(x)
  def diff(s: Set, t: Set): Set = (x: Int) => s(x) && !t(x)
  def filter(s: Set, p: Int => Boolean): Set = (x: Int) => s(x) && p(x)

  val bound = 1000
  def forall(s: Set, p: Int => Boolean): Boolean = {
    def iter(a: Int): Boolean = {
      if (a > bound) true
      else if (s(a) && !p(a)) false
      else iter(a + 1)
    }
    iter(-bound)
  }
  def exists(s: Set, p: Int => Boolean): Boolean = !forall(s, (x: Int) => !p(x))

  def map(s: Set, f: Int => Int): Set = (x: Int) => {
    def iter(a: Int): Boolean = {
      if (a > bound) false
      else if (f(a) == x && s(a)) true
      else iter(a + 1)
    }
    iter(-bound)
  }

  def toString(s: Set): String = {
    val xs = for (i <- -bound to bound if contains(s, i)) yield i
    xs.mkString("{", ",", "}")
  }

  def printSet(s: Set) {
    println(toString(s))
  }
}
